### ISSUE TITLE
Running builds from inside a nix-shell breaks them

### ISSUE THREAD

**Issue by copumpkin** [NixOS/nix#1802]
I've noticed this on and off for months now and finally decided to file a bug. I'm root, although I don't know if that matters.

This applies to version 1.12pre5788_e3013543 but has been happening forever.

Let's say I `nix-shell -p python`. The point is to be inside a `nix-shell`. Now, let's force a trivial build:

```
[nix-shell:~]$ nix-build -E '(import <nixpkgs> {}).hello.overrideAttrs (_: { name = "foo"; })'
these derivations will be built:
  /nix/store/kli2n1nkqskg482cfg1v9gapn70n6m06-foo.drv
these paths will be fetched (0.69 MiB download, 0.69 MiB unpacked):
  /nix/store/3x7dwzq014bblazs7kq20p9hyzz0qh8g-hello-2.10.tar.gz
copying path '/nix/store/3x7dwzq014bblazs7kq20p9hyzz0qh8g-hello-2.10.tar.gz' from 'https://cache.nixos.org'...
building '/nix/store/kli2n1nkqskg482cfg1v9gapn70n6m06-foo.drv'...
/nix/store/066fqdvgfnzdr27i1ffz7cghf8r89glg-stdenv/setup: line 583: /run/user/0/nix-build-foo.drv-0/env-vars: Permission denied
unpacking sources
/nix/store/066fqdvgfnzdr27i1ffz7cghf8r89glg-stdenv/setup: line 583: /run/user/0/nix-build-foo.drv-0/env-vars: Permission denied
unpacking source archive /nix/store/3x7dwzq014bblazs7kq20p9hyzz0qh8g-hello-2.10.tar.gz
source root is hello-2.10
setting SOURCE_DATE_EPOCH to timestamp 1416139241 of file hello-2.10/ChangeLog
patching sources
/nix/store/066fqdvgfnzdr27i1ffz7cghf8r89glg-stdenv/setup: line 583: /run/user/0/nix-build-foo.drv-0/env-vars: Permission denied
configuring
/nix/store/066fqdvgfnzdr27i1ffz7cghf8r89glg-stdenv/setup: line 583: /run/user/0/nix-build-foo.drv-0/env-vars: Permission denied
configure flags: --disable-dependency-tracking --prefix=/nix/store/6g80b0f2ndc8av04rs2vl2j82rygas9k-foo
./configure: line 1792: cd: /run/user/0/nix-build-foo.drv-0/hello-2.10: Not a directory
configure: error: working directory cannot be determined
builder for '/nix/store/kli2n1nkqskg482cfg1v9gapn70n6m06-foo.drv' failed with exit code 1
error: build of '/nix/store/kli2n1nkqskg482cfg1v9gapn70n6m06-foo.drv' failed
```

Drop out of the `nix-shell` and the same build will work fine.

Now let's get into the `nix-shell` again and clear the environment first:

```
[nix-shell:~]$ env -i NIX_PATH=$NIX_PATH nix-build -E '(import <nixpkgs> {}).hello.overrideAttrs (_: { name = "bar"; })'
these derivations will be built:
  /nix/store/xlgmgva578y67y26js00g03wwbdchl87-bar.drv
building path(s) ‘/nix/store/n9jx8jn139cxxjh30801fx1ydn76cfyn-bar’
unpacking sources
unpacking source archive /nix/store/3x7dwzq014bblazs7kq20p9hyzz0qh8g-hello-2.10.tar.gz
source root is hello-2.10
setting SOURCE_DATE_EPOCH to timestamp 1416139241 of file hello-2.10/ChangeLog
patching sources
configuring
configure flags: --disable-dependency-tracking --prefix=/nix/store/n9jx8jn139cxxjh30801fx1ydn76cfyn-bar
checking for a BSD-compatible install... /nix/store/r90xqqmd36fv3s53bf92s3vxhdnbwfn7-coreutils-8.28/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /nix/store/r90xqqmd36fv3s53bf92s3vxhdnbwfn7-coreutils-8.28/bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking whether gcc understands -c and -o together... yes
checking for style of include used by make... GNU
checking dependency style of gcc... none
checking how to run the C preprocessor... gcc -E
checking for grep that handles long lines and -e... /nix/store/wia9b594w2ghzhpcjdi9mnlf8j24nm3i-gnugrep-3.1/bin/grep
checking for egrep... /nix/store/wia9b594w2ghzhpcjdi9mnlf8j24nm3i-gnugrep-3.1/bin/grep -E
checking for Minix Amsterdam compiler... no
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
[...]
/nix/store/n9jx8jn139cxxjh30801fx1ydn76cfyn-bar
```

So some environment variable defined inside the `nix-shell` environment is preventing further nix builds from working. This is on NixOS, with `nix-daemon` running in the background. I haven't narrowed down what environment variables make it unhappy.

cc @edolstra

**Comment by edolstra** [comment]
This is probably due to `TMPDIR`.

**Comment by copumpkin** [comment]
Indeed! Wrapping with `env -u TMPDIR` fixes it. What's a good solution? It's not clear to me what a better solution would be.

**Comment by dtzWill** [comment]
FWIW I can't seem to reproduce this on latest Nix 2.0 (basically currently master)--can you confirm it's still a problem?

My shell normally doesn't have TMPDIR set, but that doesn't seem to matter (setting it to something first doesn't break things AFAICT?).

Nesting nix-shell's doesn't break it either, hehe.

Oh (!) I see you're root (/run/user/0), which is odd.  When entering nix-shell as root I end up with TMPDIR=/tmp.  I can't seem to convince things to end up with /run/user/0 after a few attempts, perhaps that points to a different/related issue?

**Comment by copumpkin** [comment]
If it helps, it's trivially reproducible on an EC2 NixOS 17.09 AMI. Not sure why my `TMPDIR` differs from yours. Are you on NixOS?

**Comment by TeofilC** [comment]
I've run into this recently using the `nixos/nix:2.13.1` docker image.
Nix builds run from inside a `nix develop` seem to be broken as `TMPDIR` is set

**Comment by dzmitry-lahoda** [comment]
I am on PopOS Ubuntu 22.04 latest updates and do

```
nix-shell -p python3
```

and do in shell
```
nix-build -E '(import <nixpkgs> {}).hello.overrideAttrs (_: { name = "foo"; })'
```

it success with 
```
f7b52djrx01zgzcw-foo
shrinking /nix/store/d5zgfvk7wyfan76sf7b52djrx01zgzcw-foo/bin/hello
checking for references to /build/ in /nix/store/d5zgfvk7wyfan76sf7b52djrx01zgzcw-foo...
gzipping man pages under /nix/store/d5zgfvk7wyfan76sf7b52djrx01zgzcw-foo/share/man/
patching script interpreter paths in /nix/store/d5zgfvk7wyfan76sf7b52djrx01zgzcw-foo
stripping (with command strip and flags -S -p) in  /nix/store/d5zgfvk7wyfan76sf7b52djrx01zgzcw-foo/bin
/nix/store/d5zgfvk7wyfan76sf7b52djrx01zgzcw-foo
```

System:
```
bash-5.2$ uname -a
Linux pop-os 6.8.0-76060800daily20240311-generic #202403110203~1714077665~22.04~4c8e9a0 SMP PREEMPT_DYNAMIC Thu A x86_64 GNU/Linux

bash-5.2$ nix --version --verbose
nix (Nix) 2.18.1
```

In shell I have
```
[nix-shell:~/github.com/Layer-N/nord]$ env | grep TMP
TMPDIR=/tmp/nix-shell.tsYKWj
TMP=/tmp/nix-shell.tsYKWj
````

when I set TMP to garbage

```
[nix-shell:~/github.com/Layer-N/nord]$ env | grep TMP
TMPDIR=asddasd
TMP=asddasd

[nix-shell:~/github.com/Layer-N/nord]$  nix-build -E '(import <nixpkgs> {}).hello.overrideAttrs (_: { name = "foo"; })'
error: not an absolute path: 'asddasd'

```

when i set read only dir
```
[nix-shell:~/github.com/Layer-N/nord]$ export TMP=/

[nix-shell:~/github.com/Layer-N/nord]$ export TMPDIR=/

[nix-shell:~/github.com/Layer-N/nord]$  nix-build -E '(import <nixpkgs> {}).hello.overrideAttrs (_: { name = "foo"; })'
error: creating directory '//nix-build-23046-0': Permission denied
```

**Comment by dzmitry-lahoda** [comment]
@TeofilC in docker it is the case that you can have issues:

1. TMP is too small free space
2. nor TMP/TMPDIR set
3. not USER set

So something is not set I guess.

> sandbox = false

It does not seems good solution as it allows to do IP requests from build, not only read any files.

**Comment by dzmitry-lahoda** [comment]
```
docker run -it nixos/nix:2.18.1 nix-shell -p python3
```

I am in docker, run
```
nix-build -E '(import <nixpkgs> {}).hello.overrideAttrs (_: { name = "foo"; })'
```

success
```
patching script interpreter paths in /nix/store/cs5da3l6sm0gpfib1knb2zcjx4ykqdxr-foo
stripping (with command strip and flags -S -p) in  /nix/store/cs5da3l6sm0gpfib1knb2zcjx4ykqdxr-foo/bin
/nix/store/cs5da3l6sm0gpfib1knb2zcjx4ykqdxr-foo

[nix-shell:/]# ls -a
.  ..  .dockerenv  bin	dev  etc  nix  proc  result  root  sys	tmp  usr  var

[nix-shell:/]# 

```

**Comment by dzmitry-lahoda** [comment]
@fricklerhandwerk I suggest closing issue:

1. Issue is not reproducible or fixed.
2. Nix on Ubuntu is UX is good
3. Docker is good
4. No reprodusible steps exist on latest nix.
5. Foreign dockers (not nixos) is up to people to setup properly (nix likely will report good errors inside on build)

**Comment by TeofilC** [comment]
Here is a reproducer @dzmitry-lahoda:

It occurs with the latest version of the official nix docker image.

```
docker run -it --rm nixos/nix:2.22.1
nix develop --extra-experimental-features nix-command\ flakes nixpkgs#tree
nix build --extra-experimental-features nix-command\ flakes nixpkgs#hello --rebuild
```

You get the following output:
```
       > patching script interpreter paths in ./configure
       > ./configure: interpreter directive changed from "#! /bin/sh" to "/nix/store/h3bhzvz9ipglcybbcvkxvm4vg9lwvqg4-bash-5.2p26/bin/sh"
       > configure flags: --disable-dependency-tracking --prefix=/nix/store/fhilj8ncq93z1vlspdp0w7wv27qa2mn4-hello-2.12.1
       > ./configure: line 2119: cd: /tmp/nix-shell.8meG9m/nix-build-hello-2.12.1.drv-0/hello-2.12.1: Not a directory
       > configure: error: working directory cannot be determined
       For full logs, run 'nix log /nix/store/w0hjp1fnvyf1z30s29fj4im8c7bxqrql-hello-2.12.1.drv'.
```

**Comment by fricklerhandwerk** [comment]
@TeofilC this would mean the issue does not actually exist as stated?

**Comment by dzmitry-lahoda** [comment]
@TeofilC thank you for log lines. so issue now in `nix develop`, not `nix-shell`. i also got same error. 

I did not found `working directory cannot be determined` issue on Nix, so issues may be was fixed in `nix-shell`, but not in `nix develop`. 

```
  > ./configure: line 2119: cd: /tmp/nix-shell.8meG9m/nix-build-hello-2.12.1.drv-0/hello-2.12.1: Not a directory
```

Seems directory was not created and same fix to be applied as it was to shell.  Would be awesome I can patch it, but my eyes bleeding looking at C++ :(
